api4/user.go

```diff
@@ -429,6 +434,123 @@ func setDefaultProfileImage(c *Context, w http.ResponseWriter, r *http.Request)
 	ReturnStatusOK(w)
 }
 
+func getIconImage(c *Context, w http.ResponseWriter, r *http.Request) {
+	c.RequireUserId()
+	if c.Err != nil {
+		return
+	}
+
+	canSee, err := c.App.UserCanSeeOtherUser(c.App.Session.UserId, c.Params.UserId)
+	if err != nil {
+		c.Err = err
+		return
+	}
+
+	if !canSee {
+		c.SetPermissionError(model.PERMISSION_VIEW_MEMBERS)
+		return
+	}
+
+	user, err := c.App.GetUser(c.Params.UserId)
+	if err != nil {
+		c.Err = err
+		return
+	}
+
+	etag := strconv.FormatInt(user.LastPictureUpdate, 10)
+	if c.HandleEtag(etag, "Get Icon Image", w, r) {
+		return
+	}
+
+	img, readFailed, err := c.App.GetIconImage(user)
+	if err != nil {
+		c.Err = err
+		return
+	}
+
+	if readFailed {
+		w.Header().Set("Cache-Control", fmt.Sprintf("max-age=%v, public", 5*60)) // 5 mins
+	} else {
+		w.Header().Set("Cache-Control", fmt.Sprintf("max-age=%v, public", 24*60*60)) // 24 hrs
+		w.Header().Set(model.HEADER_ETAG_SERVER, etag)
+	}
+
+	w.Header().Set("Content-Type", "image/svg+xml")
+	w.Write(img)
+}
+
+func setIconImage(c *Context, w http.ResponseWriter, r *http.Request) {
+	defer io.Copy(ioutil.Discard, r.Body)
+
+	c.RequireUserId()
+	if c.Err != nil {
+		return
+	}
+
+	if !c.App.SessionHasPermissionToUser(c.App.Session, c.Params.UserId) {
+		c.SetPermissionError(model.PERMISSION_EDIT_OTHER_USERS)
+		return
+	}
+
+	if len(*c.App.Config().FileSettings.DriverName) == 0 {
+		c.Err = model.NewAppError("setIconImage", "api.user.user_icon_image.storage.app_error", nil, "", http.StatusNotImplemented)
+		return
+	}
+
+	if r.ContentLength > *c.App.Config().FileSettings.MaxFileSize {
+		c.Err = model.NewAppError("setIconImage", "api.user.set_icon_image.too_large.app_error", nil, "", http.StatusRequestEntityTooLarge)
+		return
+	}
+
+	if err := r.ParseMultipartForm(*c.App.Config().FileSettings.MaxFileSize); err != nil {
+		c.Err = model.NewAppError("setIconImage", "api.user.set_icon_image.parse.app_error", nil, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	m := r.MultipartForm
+	imageArray, ok := m.File["image"]
+	if !ok {
+		c.Err = model.NewAppError("setIconImage", "api.user.set_icon_image.no_file.app_error", nil, "", http.StatusBadRequest)
+		return
+	}
+
+	if len(imageArray) <= 0 {
+		c.Err = model.NewAppError("setIconImage", "api.user.set_icon_image.array.app_error", nil, "", http.StatusBadRequest)
+		return
+	}
+
+	imageData := imageArray[0]
+	if err := c.App.SetIconImage(c.Params.UserId, imageData); err != nil {
```

Comment:
Just to make sure I understand this correct:
You can call `SetIconImage` for normal, non bot users, too? It this icon used anywhere for non bot users?

Reply:
I have this doubt too

Reply:
@hanzei, @jespino -- Yes this can be called for non-bot users. However we're _only_ using it for just bots at the moment.

Reply:
Even though we are just using it for bots, should we restrict it only for bots ? Overriding the user icon may interfere with user status changing icons and not work as expected?

Reply:
> should we restrict it only for bots

:+1: for this. Maybe also the API routes should be `/bot/icon` instead of `user/icon`. 2/5

Reply:
Updated routes to `/bot/<botuserid>/icon` and storing icons under `/bots/<botuserid>/icon.svg`